File: medtech_bpa/medtech_bpa/custom_scripts/stock_entry/stock_entry.js

frappe.ui.form.on("Stock Entry", {
	setup: (frm) => {
		frm.set_query('production_plan', () => {
			return {
				filters: {
					'status' :'In Process' ,
					'company': frm.doc.company
				}
			};
		});
		frm.set_query('work_order', () => {
			return {
				filters: {
					'company': frm.doc.company,
					'production_plan':frm.doc.production_plan,
					'production_item':frm.doc.item
				}
			};
		});
	},
	production_plan:function(frm){
		if(frm.doc.production_plan){
			frappe.call({
				"method":"medtech_bpa.medtech_bpa.custom_scripts.stock_entry.stock_entry.get_items_from_production_plan",
				args:{
					'production_plan':frm.doc.production_plan
				},
				callback:function(r){
					if(r.message){
						cur_frm.fields_dict['item'].get_query = function(doc, cdt, cdn) {
			return {
				filters: [
					['Item','name', 'in', r.message]
				]
			}
		}
						}
					}
							
				
			})
		}
	},
	item:function(frm){
		if(frm.doc.production_plan){
			if(frm.doc.item){
				frappe.call({
				"method":"medtech_bpa.medtech_bpa.custom_scripts.stock_entry.stock_entry.get_work_orders",
				"args":{
					'production_plan':frm.doc.production_plan,
					'item':frm.doc.item
				},
				callback:function(r){
					if(r.message){
						frm.set_value('work_order',r.message)
						// frm.set_df_property("work_order","read_only",1)
					}else{
						frappe.throw(__("Error: No Work Order Created For Item {0} and Production Plan {1} ",
					[frm.doc.item,frm.doc.production_plan]
				));
					}
				}
			})
			}
			
		}else{
			frappe.throw("Please Select Production Plan First")
		}

	},
});


frappe.ui.form.on("Stock Entry Detail", {
	t_warehouse :function(frm,cdt,cdn){
		var result= ''
		frappe.call({
			method: "frappe.client.get_value",
			args: {
				doctype: "MedTech Settings",
				fieldname: ['rm_warehouse']
			},
			callback: function(r) {
				var row = locals[cdt][cdn]
				if(r.message) {
					if(r.message.rm_warehouse == row.t_warehouse && !frm.doc.purchase_receipt && frm.doc.purpose == 'Material Transfer'){
						frappe.msgprint("Please Enter Visual Inspection Report No")

						// frm.set_df_property('purchase_receipt' , 'reqd', 1)

					}
				}
			}
		});
	}
});

File: medtech_bpa/medtech_bpa/custom_scripts/stock_entry/stock_entry.py
from __future__ import unicode_literals
from frappe.model.document import Document
import frappe
from frappe.utils import flt,today
from frappe import _


def after_insert(doc, method):
	get_warehouse = frappe.get_single('MedTech Settings')
	rm_warehouse_list = [item.warehouse for item in get_warehouse.rm_warehouse_list]
	rm_child_warehouse_list = []
	for warehouse in rm_warehouse_list:
		child_warehouse = frappe.db.get_descendants('Warehouse', warehouse)
		if child_warehouse:
			for row in child_warehouse:
				rm_child_warehouse_list.append(row)
		else:
			rm_child_warehouse_list.append(warehouse)
	qc_warehouse_list = [item.warehouse for item in get_warehouse.qc_warehouse_list]
	qc_child_warehouse_list = []
	for warehouse in qc_warehouse_list:
		child_warehouse = frappe.db.get_descendants('Warehouse', warehouse)
		if child_warehouse:
			for row in child_warehouse:
				qc_child_warehouse_list.append(row)
		else:
			qc_child_warehouse_list.append(warehouse)
	supplier = frappe.db.get_value('Purchase Receipt', doc.purchase_receipt, 'supplier')
	qc_disable_items = get_qc_disable_items(supplier)
	if doc.stock_entry_type == 'Material Transfer':
		if doc.items:
			for item in doc.items:
				if item.item_code not in qc_disable_items and item.t_warehouse in rm_child_warehouse_list and  item.s_warehouse in qc_child_warehouse_list:
						doc.inspection_required = 1


def on_submit(doc, method):
	update_qc_reference_on_vir(doc)


def get_qc_disable_items(supplier):
	query = '''SELECT qd.item_code  from `tabQC Disable Supplier` qds join `tabQC Disable` qd 
		on qds.parent = qd.name  where qds.supplier = '{0}' '''.format(supplier)
	qc_disable_items = frappe.db.sql(query, as_dict=1)
	qc_disable_items = [ item.get('item_code') for item in qc_disable_items]
	return qc_disable_items

@frappe.whitelist()
def get_work_orders(production_plan,item):
	work_order = frappe.db.get_value("Work Order",{'production_plan':production_plan,'production_item':item},'name')
	return work_order

@frappe.whitelist()
def get_items_from_production_plan(production_plan):
	items = frappe.get_list("Production Plan Item",{'parent':production_plan},'item_code')
	item_list = [item.item_code for item in items]
	return item_list


def update_qc_reference_on_vir(doc):
	get_warehouse = frappe.get_single('MedTech Settings')
	rm_warehouse_list = [item.warehouse for item in get_warehouse.rm_warehouse_list]
	rm_child_warehouse_list = []
	for warehouse in rm_warehouse_list:
		child_warehouse = frappe.db.get_descendants('Warehouse', warehouse)
		if child_warehouse:
			for row in child_warehouse:
				rm_child_warehouse_list.append(row)
		else:
			rm_child_warehouse_list.append(warehouse)
	qc_warehouse_list = [item.warehouse for item in get_warehouse.qc_warehouse_list]
	qc_child_warehouse_list = []
	for warehouse in qc_warehouse_list:
		child_warehouse = frappe.db.get_descendants('Warehouse', warehouse)
		if child_warehouse:
			for row in child_warehouse:
				qc_child_warehouse_list.append(row)
		else:
			qc_child_warehouse_list.append(warehouse)
	if doc.stock_entry_type == 'Material Transfer':
		if doc.items:
			for item in doc.items:
				if item.t_warehouse in rm_child_warehouse_list and  item.s_warehouse in qc_child_warehouse_list and item.quality_inspection:
					rejected_qty = frappe.db.get_value('Quality Inspection', {'name' : item.quality_inspection}, 'rejected_quantity') or 0

					if rejected_qty > 0:
						# create mr for rejected warehouse
						current_date = frappe.utils.today()
						stock_entry = frappe.new_doc("Stock Entry")
						if stock_entry:
							stock_entry.posting_date = current_date
							stock_entry.stock_entry_type = "Rejected Material Transfer"
							stock_entry.purchase_receipt = doc.purchase_receipt
							stock_entry.append("items",{
								'item_code': item.get('item_code'),
								'item_name': item.get('item_name'),
								'item_group':item.get('item_group'),
								'description': item.get('description'),
								'uom': item.get('uom'),
								'qty': rejected_qty,
								's_warehouse': item.t_warehouse,
								't_warehouse': get_warehouse.rejected_warehouse,
								'basic_rate' : item.get('rate')
							})
							stock_entry.save(ignore_permissions = True)
							stock_entry.submit()
							frappe.db.commit()	

					frappe.db.sql("Update `tabPurchase Receipt Item` set quality_inspection = '{0}', custom_rejected_qty = '{1}' where parent = '{2}' and item_code = '{3}'".format(item.quality_inspection, rejected_qty, doc.purchase_receipt, item.item_code))
					frappe.db.commit()